prov/util/src/util_cq.c

```diff
@@ -250,62 +250,13 @@ static void util_cq_read_tagged(void **dst, void *src)
 ssize_t ofi_cq_readfrom(struct fid_cq *cq_fid, void *buf, size_t count,
 			fi_addr_t *src_addr)
 {
-	struct fi_cq_tagged_entry *entry;
-	struct util_cq_aux_entry *aux_entry;
 	struct util_cq *cq;
-	ssize_t i;
 
 	cq = container_of(cq_fid, struct util_cq, cq_fid);
 
 	cq->progress(cq);
-	ofi_genlock_lock(&cq->cq_lock);
-	if (ofi_cirque_isempty(cq->cirq)) {
-		i = -FI_EAGAIN;
-		goto out;
-	}
 
-	if (count > ofi_cirque_usedcnt(cq->cirq))
-		count = ofi_cirque_usedcnt(cq->cirq);
-
-	for (i = 0; i < (ssize_t) count; i++) {
-		entry = ofi_cirque_head(cq->cirq);
-		if (!(entry->flags & UTIL_FLAG_AUX)) {
-			if (src_addr && cq->src)
-				src_addr[i] = cq->src[ofi_cirque_rindex(cq->cirq)];
-			cq->read_entry(&buf, entry);
-			ofi_cirque_discard(cq->cirq);
-		} else {
-			assert(!slist_empty(&cq->aux_queue));
-			aux_entry = container_of(cq->aux_queue.head,
-						 struct util_cq_aux_entry,
-						 list_entry);
-			assert(aux_entry->cq_slot == entry);
-			if (aux_entry->comp.err) {
-				if (!i)
-					i = -FI_EAVAIL;
-				break;
-			}
-
-			if (src_addr && cq->src)
-				src_addr[i] = aux_entry->src;
-			cq->read_entry(&buf, &aux_entry->comp);
-			slist_remove_head(&cq->aux_queue);
-			free(aux_entry);
-
-			if (slist_empty(&cq->aux_queue)) {
-				ofi_cirque_discard(cq->cirq);
-			} else {
-				aux_entry = container_of(cq->aux_queue.head,
-							struct util_cq_aux_entry,
-							list_entry);
-				if (aux_entry->cq_slot != ofi_cirque_head(cq->cirq))
-					ofi_cirque_discard(cq->cirq);
-			}
-		}
-	}
-out:
-	ofi_genlock_unlock(&cq->cq_lock);
-	return i;
+	return ofi_cq_read_cq(cq, buf, count, src_addr);
```

Comment:
Why is this change being made?  What is the motivation?

Reply:
In our peer provider implementation, we want to make efa's `fi_cq_readfrom` exit early after we do `fi_cq_read(shm_cq, NULL, 0)` and the cqe is written, so we do not have to progress efa cq, which is more expensive than progressing shm cq. This will reduce some overhead. So after introducing this split, we can do the following in efa's cq_readfrom:

```
static ssize_t efa_rdm_cq_readfrom(struct fid_cq *cq_fid, void *buf, size_t count, fi_addr_t *src_addr)
{
	struct efa_rdm_cq *cq;
	ssize_t ret;

	cq = container_of(cq_fid, struct efa_rdm_cq, util_cq.cq_fid.fid);

	if (cq->shm_cq)
		fi_cq_read(cq->shm_cq, NULL, 0);

	ret = ofi_cq_flush_cirque(&cq->util_cq, buf, count, src_addr);

	if (ret > 0)
		return ret;

	return ofi_cq_readfrom(&cq->util_cq.cq_fid, buf, count, src_addr);
}
```

Reply:
Thank you, added!